external_libraries/mpi_python/python_interface.cpp

```diff
@@ -23,20 +38,27 @@ PYBIND11_MODULE(mpipython, m)
     RankFuncType FRank = &PythonMPI::rank;
     SizeFuncType FSize = &PythonMPI::size;
 
-    class_<PythonMPI>(m,"PythonMPI") 
+    // note that for the functions returning a vector the conversion to a python-list is automatically
+    // done by pybind, see https://github.com/pybind/pybind11/blob/master/docs/advanced/cast/stl.rst
+    std::vector<int> (PythonMPI::*gather_int)(PythonMPIComm&, const int, const int) = &PythonMPI::gather;
```

Comment:
why do you use this mechanism here? it looks to me that the old way was actually cleaner...

i am speaking for example of 

.def("gather",&PythonMPI::gather<double>)

Reply:
in any case i think that making the conversion explicit could be actually nicer.

for example one could do in python

    my_vec = gather....
    my_list = list( my_vec )
    my_set = set( my_vec )

i think that all of this would work automatically in python and it goes in the same direction as #2542

Reply:
I see, I will try to change it

The thing regarding the conversion, I am not sure if pybind allows to return an std::vector in python, the conversion to the list is done automatically

Otherwise I would have to return a list again, which imo is the uglier solution

Reply:
Just tried both things:
- regarding the function pointers I use for exposing in python: I need those bcs the template argument is the same in both cases, therefore I cannot use it directly
- regarding the return type: Pybind throws the following error if I try to give back the vector directly to python:
~~~
TypeError: Unable to convert function return value to a Python type! The signature was
	(self: mpipython.PythonMPI, arg0: mpipython.PythonMPIComm, arg1: int, arg2: int) -> std::vector<int, std::allocator<int> >
~~~

Reply:
@RiccardoRossi can I go ahead then?

I would like to add some more functions/capabilities in a next PR

Reply:
Hi @philbucher 
      i think that the problem is that you didn't expose the type std::vector< std::vector< ... > > to python. 

if you give back a py::list of std::vectors, does that work?

sorry for insisting but i think that the ability to transform from/to a list or set is very important, and i really would not like to lose that ability. for example with the current version can we iterate over the return type?

something like

     for item in returned_vector_of_vectors:
          for value in item:
               print(item)

?

Reply:
mmm, i just realized that this capability was not exposed before. Unfortunately this also implies that there is no test whatsoever for this :-(:-( ... which brings me to ask adding for at least a minimum coverage, nothing fancy but just to check it works